text/0068-entity-tags.md

```diff
@@ -0,0 +1,296 @@
+# Entity tags
```

Comment:
I find "tags" to be a confusing name for this feature. When I read "tag", my first thought is that entities just have some set of strings attached, but tags was proposed are really extra attributes. I think it would be much less confusing if we could find a way to fold this into the existing notion of records.

Reply:
RFCs 66 and 27, linked from this RFC are proposals to reuse the existing notion of records either as "open records" (records have a default type that is associated with unspecified attributes), or map types (a different type given to records where all attributes are unspecified), respectively. Both of these RFCs have been rejected due to concerns about undecidable or inefficient logical encodings, precluding effective analyzability. By contrast, this RFC _is_ analyzable efficiently. Moreover, it turns out that those prior proposals were all motivated by wanting tags. The more general language feature (first-class records with default attribute types) might be pleasing at some level, but was not at all necessary for the customer use-case.

Reply:
Starting from the idea that AWS style tags are a good feature, then I like this RFCs proposed implementation of tags in Cedar.

I suppose it must be a good feature to the extent that a customer wants to use it, but  I'm not fully convinced that beyond that they're a good general purpose feature customers outside the AWS ecosystem will want.

Reply:
What do you mean by "outside of the AWS ecosystem?" We have non-AWS customers who want tags. See https://github.com/cedar-policy/cedar-examples/tree/main/cedar-example-use-cases/tags_n_roles as an example of policies that use tags following one of the listed workarounds, which would be improved by this RFC (no need to change schemas on the fly, as you define new tags or roles). This example is based on a customer application.

Reply:
+ 1 for changing the name of this concept.  How about "Open Entities"?

Reply:
I tried to come up with an alternative name but didn't manage to. To me, an "open entity" is an entity with the feature proposed here, i.e., a thing called an _X_ which has type `{ ?: T }`. The presence of an attribute of that type makes the entity open, but the attribute _itself_ is not an open entity. This RFC calls an entity attribute of type `{ ?: T }` "entity tags".

If you can come up with a replacement name _X_ whereby you mostly search-and-replace "entity tags" with _X_ and it works then I'm happy to entertain it!

Reply:
Tags have a lot of similar meanings to what we want to achieve here without being affiliated to AWS. It always means a way to organize objects and we aimed here at passing those organizing metadata for policies evaluation.
Tags exists in devops world (Github, Gitlab, JIRA), IAC world (Terraform, Chef, Ansible), SaaS world (Salesforce), other security products... I am pretty sure I can go on :)

Reply:
> It always means a way to organize objects and we aimed here at passing those organizing metadata for policies evaluation.

I agree that tags are a common concept but, in my experience at least, "tag" usually means a single piece of meta-data that is attached to an object.  I haven't seen the word "tags" used to denote a set of key-value pairs that are part of the object's data, which is what we have here.   

> If you can come up with a replacement name X whereby you mostly search-and-replace "entity tags" with X and it works then I'm happy to entertain it!

How about "embedded maps" or "entity maps"?  The word "map" conveys what the underlying datastructure is a set of key-value pairs whose keys aren't known in advance, and the word "embedded" or "entity" signals that these aren't regular maps / first-class values.

Reply:
I like "embedded maps" but: The keys to the maps are always strings, so they are not as general as maps. Will that be confusing?

Another thought is "embedded attribute maps" to emphasize that such a map is from attributes to values (or, that it must always be stored in an attribute).

Reply:
I like "embedded attribute maps" too.